File: src/ru/nsu/ccfit/zuev/osu/menu/ModMenu.java
package ru.nsu.ccfit.zuev.osu.menu;

import com.edlplan.ui.fragment.ModSettingsMenu;
import com.reco1l.ibancho.RoomAPI;
import com.reco1l.osu.data.BeatmapInfo;
import com.reco1l.osu.Execution;
import com.reco1l.osu.multiplayer.Multiplayer;
import com.reco1l.ibancho.data.RoomMods;
import com.reco1l.osu.multiplayer.RoomScene;

import com.rian.osu.GameMode;
import com.rian.osu.beatmap.parser.BeatmapParser;
import com.rian.osu.difficulty.BeatmapDifficultyCalculator;
import com.rian.osu.mods.*;
import com.rian.osu.utils.ModHashMap;

import org.anddev.andengine.entity.primitive.Rectangle;
import org.anddev.andengine.entity.scene.Scene;
import org.anddev.andengine.entity.text.ChangeableText;
import org.anddev.andengine.input.touch.TouchEvent;
import org.anddev.andengine.opengl.texture.region.TextureRegion;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CancellationException;

import org.anddev.andengine.util.MathUtils;

import kotlinx.coroutines.Job;
import ru.nsu.ccfit.zuev.osu.*;
import ru.nsu.ccfit.zuev.osu.game.GameHelper;
import ru.nsu.ccfit.zuev.osu.game.mods.IModSwitcher;
import ru.nsu.ccfit.zuev.osu.game.mods.ModButton;
import ru.nsu.ccfit.zuev.osu.helper.StringTable;
import ru.nsu.ccfit.zuev.osu.helper.TextButton;

public class ModMenu implements IModSwitcher {
    private static final ModMenu instance = new ModMenu();
    private Scene scene = null, parent;
    private ModHashMap enabledMods = new ModHashMap();
    private ChangeableText multiplierText;
    private BeatmapInfo selectedBeatmap;
    private final Map<Mod, ModButton> modButtons = new HashMap<>();
    private boolean enableNCWhenSpeedChange = false;
    private Job calculationJob;
    private ModSettingsMenu menu;

    private ModMenu() {}
    
    public static ModMenu getInstance() {
        return instance;
    }

    public void show(Scene scene, BeatmapInfo selectedBeatmap) {
        parent = scene;
        setSelectedTrack(selectedBeatmap);
        scene.setChildScene(getScene(), false, true, true);
        if (menu == null) {
            menu = new ModSettingsMenu();
        }

        Execution.mainThread(menu::show);
        update();
    }

    public void update()
    {
        // Ensure selected mods are visually selected
        synchronized (modButtons) {
            for (var key : modButtons.keySet()) {
                var button = modButtons.get(key);

                if (button != null)
                    button.setEnabled(enabledMods.contains(key));
            }

            // Updating multiplier text just in case
            changeMultiplierText();
        }
    }

    public void setMods(RoomMods mods, boolean isFreeMods, boolean allowForceDifficultyStatistics)
    {
        var modMap = mods.map;

        if (isFreeMods) {
            for (var mod : modMap.values()) {
                if (!mod.isValidForMultiplayerAsFreeMod()) {
                    enabledMods.put(mod);
                }
            }
        } else {
            enabledMods = modMap;
        }

        if (!isFreeMods || !allowForceDifficultyStatistics) {
            var difficultyAdjust = modMap.ofType(ModDifficultyAdjust.class);

            if (difficultyAdjust != null) {
                enabledMods.put(difficultyAdjust);
            }
        }

        if (!Multiplayer.isRoomHost() && (modMap.contains(ModDoubleTime.class) || modMap.contains(ModNightCore.class))) {
            var doubleTime = new ModDoubleTime();
            var nightCore = new ModNightCore();

            enabledMods.remove(Config.isUseNightcoreOnMultiplayer() ? doubleTime : nightCore);
            enabledMods.put(Config.isUseNightcoreOnMultiplayer() ? nightCore : doubleTime);
        }

        update();
    }

    public void hide() {
        hide(true);
    }

    public void hide(boolean updatePlayerMods) {
        if (parent != null) {
            parent.clearChildScene();
            parent = null;
        }
//        InGameSettingMenu.Companion.getInstance().dismiss();
        if (menu != null) {
            menu.dismiss();
        }

        if (Multiplayer.isConnected())
        {
            RoomScene.isWaitingForModsChange = true;

            var string = enabledMods.toString();

            // The room mods are the same as the host mods
            if (Multiplayer.isRoomHost()) {
                RoomAPI.setRoomMods(string);
            } else if (updatePlayerMods) {
                RoomAPI.setPlayerMods(string);
            } else {
                RoomScene.isWaitingForModsChange = false;
            }
        }
    }

    public void hideByFrag() {
        if (parent != null) {
            parent.clearChildScene();
            parent = null;
        }

        menu = null;
    }

    private void addButton(int x, int y, Mod mod) {
        if (!(mod instanceof IModUserSelectable selectableMod)) {
            throw new IllegalArgumentException("Mod must implement IModUserSelectable");
        }

        var button = new ModButton(x, y, selectableMod, this);

        button.setEnabled(this.enabledMods.contains(mod));
        scene.attachChild(button);
        scene.registerTouchArea(button);

        modButtons.put(mod, button);
    }

    public void init() {
        cancelCalculationJob();
        calculationJob = null;

        modButtons.clear();
        scene = new Scene();
        scene.setBackgroundEnabled(false);
        final Rectangle bg = new Rectangle(0, 0, Config.getRES_WIDTH(),
                Config.getRES_HEIGHT());
        bg.setColor(0, 0, 0, 0.7f);
        scene.attachChild(bg);

        multiplierText = new ChangeableText(0, Utils.toRes(50),
                ResourceManager.getInstance().getFont("CaptionFont"),
                StringTable.format(com.osudroid.resources.R.string.menu_mod_multiplier, 1f));
        multiplierText.setScale(1.2f);
        scene.attachChild(multiplierText);

        menu = new ModSettingsMenu();

        changeMultiplierText();

        final int offset = 100;
        final int offsetGrowth = 130;
        final TextureRegion button = ResourceManager.getInstance().getTexture("selection-mod-easy");

        var clickShortSound = ResourceManager.getInstance().getSound("click-short");
        var clickShortConfirmSound = ResourceManager.getInstance().getSound("click-short-confirm");

        //line 1
        addButton(offset, Config.getRES_HEIGHT() / 2 - button.getHeight() * 3, new ModEasy());

        // Used to define the X offset of each button according to its visibility
        int factor = 1;

        addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 - button.getHeight() * 3, new ModNoFail());

        if (!Multiplayer.isMultiplayer || Multiplayer.isRoomHost())
            addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 - button.getHeight() * 3, new ModHalfTime());

        addButton(offset + offsetGrowth * factor, Config.getRES_HEIGHT() / 2 - button.getHeight() * 3, new ModReallyEasy());

        factor = 1;

        //line 2
        addButton(offset, Config.getRES_HEIGHT() / 2 - button.getHeight() / 2, new ModHardRock());

        if (!Multiplayer.isMultiplayer || Multiplayer.isRoomHost())
            addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 - button.getHeight() / 2, new ModDoubleTime());

        if (!Multiplayer.isMultiplayer || Multiplayer.isRoomHost())
            addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 - button.getHeight() / 2, new ModNightCore());

        addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 - button.getHeight() / 2, new ModHidden());
        addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 - button.getHeight() / 2, new ModTraceable());
        addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 - button.getHeight() / 2, new ModFlashlight());
        addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 - button.getHeight() / 2, new ModSuddenDeath());
        addButton(offset + offsetGrowth * factor, Config.getRES_HEIGHT() / 2 - button.getHeight() / 2, new ModPerfect());

        factor = 1;

        //line 3
        addButton(offset, Config.getRES_HEIGHT() / 2 + button.getHeight() * 2, new ModRelax());
        addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 + button.getHeight() * 2, new ModAutopilot());

        if (!Multiplayer.isMultiplayer)
            addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 + button.getHeight() * 2, new ModAuto());

        if (!Multiplayer.isMultiplayer)
            addButton(offset + offsetGrowth * factor++, Config.getRES_HEIGHT() / 2 + button.getHeight() * 2, new ModScoreV2());

        addButton(offset + offsetGrowth * factor, Config.getRES_HEIGHT() / 2 + button.getHeight() * 2, new ModPrecise());

        final TextButton resetText = new TextButton(ResourceManager
                .getInstance().getFont("CaptionFont"),
                StringTable.get(com.osudroid.resources.R.string.menu_mod_reset)) {

            boolean moved = false;
            private float dx = 0, dy = 0;

            @Override
            public boolean onAreaTouched(final TouchEvent pSceneTouchEvent,
                                         final float pTouchAreaLocalX, final float pTouchAreaLocalY) {
                if (pSceneTouchEvent.isActionDown()) {
                    moved = false;
                    dx = pTouchAreaLocalX;
                    dy = pTouchAreaLocalY;
                    return true;
                }

                if (pSceneTouchEvent.isActionUp()) {
                    if (!moved) {
                        if (clickShortConfirmSound != null) {
                            clickShortConfirmSound.play();
                        }

                        enabledMods.clear();
                        reloadMusicEffects();
                        changeMultiplierText();
                        for (ModButton btn : modButtons.values()) {
                            btn.setEnabled(false);
                        }
                    }

                    return true;
                }

                if (pSceneTouchEvent.isActionOutside()
                        || pSceneTouchEvent.isActionMove()
                        && (MathUtils.distance(dx, dy, pTouchAreaLocalX,
                        pTouchAreaLocalY) > 50)) {
                    if (!moved && clickShortSound != null) {
                        clickShortSound.play();
                    }

                    moved = true;
                }

                return false;
            }
        };

        if (!Multiplayer.isMultiplayer) {
            scene.attachChild(resetText);
            scene.registerTouchArea(resetText);
        }
        resetText.setScale(1.2f);

        final TextButton back = new TextButton(ResourceManager
                .getInstance().getFont("CaptionFont"),
                StringTable.get(com.osudroid.resources.R.string.menu_mod_back)) {

            boolean moved = false;
            private float dx = 0, dy = 0;

            @Override
            public boolean onAreaTouched(final TouchEvent pSceneTouchEvent,
                                         final float pTouchAreaLocalX, final float pTouchAreaLocalY) {
                if (pSceneTouchEvent.isActionDown()) {
                    moved = false;
                    dx = pTouchAreaLocalX;
                    dy = pTouchAreaLocalY;
                    return true;
                }

                if (pSceneTouchEvent.isActionUp() && !moved) {
                    if (clickShortConfirmSound != null) {
                        clickShortConfirmSound.play();
                    }

                    cancelCalculationJob();

                    calculationJob = Execution.async(scope -> {
                        if (selectedBeatmap == null) {
                            return;
                        }

                        try (var parser = new BeatmapParser(selectedBeatmap.getPath(), scope)) {
                            var beatmap = parser.parse(
                                true,
                                Config.getDifficultyAlgorithm() == DifficultyAlgorithm.droid ? GameMode.Droid : GameMode.Standard
                            );
                            if (beatmap == null) {
                                GlobalManager.getInstance().getSongMenu().setStarsDisplay(0);
                                return;
                            }

                            // Copy the mods to avoid concurrent modification
                            var mods = enabledMods.deepCopy().values();

                            switch (Config.getDifficultyAlgorithm()) {
                                case droid -> {
                                    var attributes = BeatmapDifficultyCalculator.calculateDroidDifficulty(
                                        beatmap, mods, scope
                                    );

                                    GlobalManager.getInstance().getSongMenu().setStarsDisplay(
                                        GameHelper.Round(attributes.starRating, 2)
                                    );
                                }

                                case standard -> {
                                    var attributes = BeatmapDifficultyCalculator.calculateStandardDifficulty(
                                        beatmap, mods, scope
                                    );

                                    GlobalManager.getInstance().getSongMenu().setStarsDisplay(
                                        GameHelper.Round(attributes.starRating, 2)
                                    );
                                }
                            }
                        }
                    });
                    hide();
                    return true;
                }

                if (pSceneTouchEvent.isActionOutside()
                        || pSceneTouchEvent.isActionMove()
                        && (MathUtils.distance(dx, dy, pTouchAreaLocalX,
                        pTouchAreaLocalY) > 50)) {
                    if (!moved && clickShortSound != null) {
                        clickShortSound.play();
                    }
                    moved = true;
                }

                return false;
            }
        };
        back.setScale(1.2f);
        back.setWidth(resetText.getWidth());
        back.setHeight(resetText.getHeight());
        back.setPosition(Config.getRES_WIDTH() - back.getWidth() - 60, Config.getRES_HEIGHT() - back.getHeight() - 30);
        back.setColor(66 / 255f, 76 / 255f, 80 / 255f);
        resetText.setPosition(Config.getRES_WIDTH() - resetText.getWidth() - 60, back.getY() - resetText.getHeight() - 20);
//		multiplierText.setPosition(back.getX() + (back.getWidth() / 2 - multiplierText.getWidth() / 2), resetText.getY() - multiplierText.getHeight() - 40);

        scene.attachChild(back);
        scene.registerTouchArea(back);

        scene.setTouchAreaBindingEnabled(true);

        update();
    }

    public Scene getScene() {
        if (scene == null) {
            init();
        }
        return scene;
    }

    public ModHashMap getEnabledMods() {
        return enabledMods;
    }

    public void changeMultiplierText() {
        GlobalManager.getInstance().getSongMenu().changeDimensionInfo(selectedBeatmap);

        float multiplier = 1;

        if (selectedBeatmap != null) {
            var difficulty = selectedBeatmap.getBeatmapDifficulty();

            for (var mod : enabledMods.values()) {
                multiplier *= mod.calculateScoreMultiplier(difficulty);
            }
        }

        multiplierText.setText(StringTable.format(com.osudroid.resources.R.string.menu_mod_multiplier, multiplier));
        multiplierText.setPosition(
                Config.getRES_WIDTH() / 2f - multiplierText.getWidth() / 2,
                multiplierText.getY());
        if (multiplier == 1) {
            multiplierText.setColor(1, 1, 1);
        } else if (multiplier < 1) {
            multiplierText.setColor(1, 150f / 255f, 0);
        } else {
            multiplierText.setColor(5 / 255f, 240 / 255f, 5 / 255f);
        }
    }

    @Override
    public boolean switchMod(IModUserSelectable selectableMod) {
        var mod = (Mod) selectableMod;
        boolean returnValue = true;

        var checkOffSound = ResourceManager.getInstance().getSound("check-off");
        var checkOnSound = ResourceManager.getInstance().getSound("check-on");

        if (enabledMods.contains(mod)) {
            enabledMods.remove(mod);
            returnValue = false;
        } else {
            enabledMods.put(mod);
        }

        if (returnValue) {
            if (checkOnSound != null) {
                checkOnSound.play();
            }
        } else {
            if (checkOffSound != null) {
                checkOffSound.play();
            }
        }

        if (mod instanceof ModRateAdjust) {
            reloadMusicEffects();
        }

        update();

        return returnValue;
    }

    public void setSelectedTrack(BeatmapInfo selectedBeatmap) {
        this.selectedBeatmap = selectedBeatmap;
        if (selectedBeatmap != null) {
            changeMultiplierText();
        }
    }

    private void reloadMusicEffects() {
        GlobalManager.getInstance().getSongMenu().updateMusicEffects();
    }

    public boolean isEnableNCWhenSpeedChange(){
        return enableNCWhenSpeedChange;
    }

    public void setEnableNCWhenSpeedChange(boolean t){
        enableNCWhenSpeedChange = t;

        GlobalManager.getInstance().getSongMenu().updateMusicEffects();
    }

    public void cancelCalculationJob() {
        if (calculationJob != null) {
            calculationJob.cancel(new CancellationException("Difficulty calculation has been cancelled."));
        }
    }

    public ModSettingsMenu getMenu() {
        return menu;
    }
}

File: src/com/osudroid/ui/v2/modmenu/ModMenu.kt
package com.osudroid.ui.v2.modmenu

import com.reco1l.andengine.*
import com.reco1l.andengine.container.*
import com.reco1l.andengine.shape.*
import com.reco1l.andengine.sprite.*
import com.reco1l.andengine.ui.*
import com.reco1l.framework.*
import com.reco1l.framework.math.*
import com.osudroid.multiplayer.api.RoomAPI.setPlayerMods
import com.osudroid.multiplayer.api.RoomAPI.setRoomMods
import com.osudroid.multiplayer.api.data.RoomMods
import com.osudroid.multiplayer.Multiplayer
import com.osudroid.ui.v2.ModsIndicator
import com.osudroid.ui.v2.StarRatingBadge
import com.osudroid.utils.updateThread
import com.reco1l.andengine.text.FontAwesomeIcon
import com.reco1l.andengine.theme.Icon
import com.reco1l.andengine.theme.Size
import com.reco1l.andengine.theme.rem
import com.reco1l.andengine.theme.srem
import com.reco1l.andengine.ui.UITextButton
import com.reco1l.toolkt.kotlin.*
import com.reco1l.toolkt.kotlin.async
import com.rian.osu.*
import com.rian.osu.beatmap.Beatmap
import com.rian.osu.beatmap.parser.*
import com.rian.osu.difficulty.BeatmapDifficultyCalculator.calculateDroidDifficulty
import com.rian.osu.difficulty.BeatmapDifficultyCalculator.calculateStandardDifficulty
import com.rian.osu.mods.*
import com.rian.osu.utils.*
import com.rian.osu.utils.ModUtils
import kotlinx.coroutines.*
import ru.nsu.ccfit.zuev.osu.*
import ru.nsu.ccfit.zuev.osu.DifficultyAlgorithm.*
import ru.nsu.ccfit.zuev.osu.helper.*
import java.util.LinkedList
import java.util.concurrent.CancellationException
import kotlin.math.*
import kotlin.reflect.KClass

object ModMenu : UIScene() {


    /**
     * List of currently enabled mods.
     */
    val enabledMods = ModHashMap()

    /**
     * List of all mod sections.
     */
    val modSections = mutableListOf<ModMenuSection>()

    /**
     * List of all mod toggles.
     */
    val modToggles = mutableListOf<ModMenuToggle>()


    private val modChangeQueue = LinkedList<Mod>()
    private val modPresetsSection: ModMenuPresetsSection

    private val customizeButton: UITextButton
    private val customizationMenu: ModCustomizationMenu
    private val selectedModsIndicator: ModsIndicator
    private val searchInput: ModMenuSearchInput

    private val rankedBadge: UIBadge
    private val arBadge: UILabeledBadge
    private val odBadge: UILabeledBadge
    private val csBadge: UILabeledBadge
    private val hpBadge: UILabeledBadge
    private val bpmBadge: UILabeledBadge
    private val starRatingBadge: StarRatingBadge
    private val scoreMultiplierBadge: UILabeledBadge

    private var parsedBeatmap: Beatmap? = null
    private var calculationJob: Job? = null


    init {
        isBackgroundEnabled = false

        customizationMenu = ModCustomizationMenu()

        attachChild(UIFillContainer().apply {
            width = Size.Full
            height = Size.Full
            orientation = Orientation.Vertical
            style = {
                backgroundColor = it.accentColor * 0.1f / 0.9f
            }

            +UIContainer().apply {
                width = Size.Full
                height = Size.Auto
                style = {
                    padding = UIEngine.current.safeArea.copy(y = 2f.srem, w = 2f.srem)
                }

                +UILinearContainer().apply {
                    orientation = Orientation.Horizontal
                    anchor = Anchor.CenterLeft
                    origin = Anchor.CenterLeft
                    style = {
                        spacing = 2f.srem
                    }

                    +UITextButton().apply {
                        text = "Back"
                        leadingIcon = FontAwesomeIcon(Icon.ArrowLeft)
                        onActionUp = {
                            ResourceManager.getInstance().getSound("click-short-confirm")?.play()
                            back()
                        }
                        onActionCancel = { ResourceManager.getInstance().getSound("click-short")?.play() }
                    }

                    customizeButton = UITextButton().apply {
                        text = "Customize"
                        isEnabled = false
                        leadingIcon = FontAwesomeIcon(Icon.Sliders)
                        onActionUp = {
                            ResourceManager.getInstance().getSound("click-short-confirm")?.play()
                            if (customizationMenu.isVisible) {
                                customizationMenu.hide()
                            } else {
                                customizationMenu.show()
                            }
                        }
                        onActionCancel = { ResourceManager.getInstance().getSound("click-short")?.play() }
                    }

                    +customizeButton

                    +UITextButton().apply {
                        text = "Clear"
                        style += {
                            color = Color4(0xFFFFBFBF)
                            backgroundColor = Color4(0xFF342121)
                        }
                        leadingIcon = FontAwesomeIcon(Icon.DeleteLeft)
                        onActionUp = {
                            ResourceManager.getInstance().getSound("click-short-confirm")?.play()
                            clear()
                        }
                        onActionCancel = { ResourceManager.getInstance().getSound("click-short")?.play() }
                    }

                    +UIScrollableContainer().apply {
                        width = 340f
                        height = Size.Auto
                        anchor = Anchor.CenterLeft
                        origin = Anchor.CenterLeft
                        scrollAxes = Axes.X
                        clipToBounds = true

                        selectedModsIndicator = ModsIndicator()
                        +selectedModsIndicator
                    }
                }

                +UIContainer().apply {
                    anchor = Anchor.CenterRight
                    origin = Anchor.CenterRight

                    searchInput = ModMenuSearchInput().apply {
                        width = 18f.rem
                        onSearchTermUpdate = { searchTerm ->
                            modSections.fastForEach { it.onSearchTermUpdate(searchTerm) }

                            if (modPresetsSection.isVisible) {
                                modPresetsSection.onSearchTermUpdate(searchTerm)
                            }
                        }
                    }

                    +searchInput

                    container {
                        anchor = Anchor.CenterRight
                        origin = Anchor.CenterRight
                        style = {
                            padding = Vec4(2f.srem, 0f)
                        }

                        +FontAwesomeIcon(Icon.MagnifyingGlass).apply {
                            style = {
                                color = it.accentColor
                            }
                        }
                    }
                }
            }

            +UIScrollableContainer().apply {
                width = Size.Full
                height = Size.Full
                scrollAxes = Axes.X

                +UILinearContainer().apply {
                    orientation = Orientation.Horizontal
                    width = Size.Auto
                    height = Size.Full
                    style = {
                        spacing = 2f.srem
                        padding = UIEngine.current.safeArea
                    }

                    modPresetsSection = ModMenuPresetsSection()
                    +modPresetsSection

                    val mods = ModUtils.allModsInstances

                    ModType.entries.forEach { type ->
                        val sectionName = StringTable.get(type.stringId)
                        val sectionToggles = mods.filter { it !is IMigratableMod && it.isUserPlayable && it.type == type }.map { ModMenuToggle(it) }

                        if (sectionToggles.isEmpty()) {
                            return@forEach
                        }

                        modToggles.addAll(sectionToggles)

                        val section = ModMenuSection(sectionName, sectionToggles)

                        +section
                        modSections.add(section)
                    }
                }
            }

            +UIContainer().apply {
                width = Size.Full
                height = Size.Auto
                style = {
                    padding = UIEngine.current.safeArea.copy(
                        y = 2f.srem,
                        w = 2f.srem + (Multiplayer.roomScene?.chat?.buttonHeight ?: 0f)
                    )
                }

                +UILinearContainer().apply {
                    orientation = Orientation.Horizontal
                    anchor = Anchor.CenterLeft
                    origin = Anchor.CenterLeft
                    style = {
                        spacing = 2f.srem
                    }

                    arBadge = labeledBadge {
                        label = "AR"
                        value = "0.00"
                    }

                    odBadge = labeledBadge {
                        label = "OD"
                        value = "0.00"
                    }

                    csBadge = labeledBadge {
                        label = "CS"
                        value = "0.00"
                    }

                    hpBadge = labeledBadge {
                        label = "HP"
                        value = "0.00"
                    }

                    bpmBadge = labeledBadge {
                        label = "BPM"
                        value = "0.0"
                    }
                }

                +UILinearContainer().apply {
                    orientation = Orientation.Horizontal
                    anchor = Anchor.CenterRight
                    origin = Anchor.CenterRight
                    style = {
                        spacing = 2f.srem
                    }

                    starRatingBadge = StarRatingBadge()
                    +starRatingBadge

                    rankedBadge = badge {
                        text = "Ranked"
                        backgroundColor = Color4(0xFF342121)
                        color = Color4(0xFF161622)
                        style = {}
                    }

                    scoreMultiplierBadge = labeledBadge {
                        label = "Score"
                        value = "1.00x"
                    }
                }
            }
        })

        // Customizations menu
        attachChild(customizationMenu)

        modPresetsSection.loadPresets()
    }


    //region Calculation

    fun cancelCalculationJob() {
        calculationJob?.cancel(CancellationException("Difficulty calculation has been cancelled."))
        calculationJob = null
    }

    private fun parseBeatmap() {
        cancelCalculationJob()

        val selectedBeatmap = GlobalManager.getInstance().selectedBeatmap ?: return

        calculationJob = async scope@{

            val difficultyAlgorithm = Config.getDifficultyAlgorithm()
            val gameMode = if (difficultyAlgorithm == droid) GameMode.Droid else GameMode.Standard
            val beatmap: Beatmap?

            if (parsedBeatmap?.md5 != selectedBeatmap.md5 || parsedBeatmap?.mode != gameMode) {
                BeatmapParser(selectedBeatmap.path, this@scope).use { parser ->
                    beatmap = parser.parse(withHitObjects = true, mode = gameMode)
                    parsedBeatmap = beatmap
                }
            } else {
                beatmap = parsedBeatmap
            }

            val songMenu = GlobalManager.getInstance().songMenu

            if (beatmap == null) {
                songMenu.setStarsDisplay(0f)
                return@scope
            }

            enabledMods.values.filterIsInstance<IModRequiresOriginalBeatmap>().fastForEach { mod ->
                ensureActive()
                mod.applyFromBeatmap(beatmap)
            }
            customizationMenu.updateComponents()

            // Copy the mods to avoid concurrent modification
            val mods = enabledMods.deepCopy().values
            val difficulty = beatmap.difficulty.clone()
            val rate = ModUtils.calculateRateWithMods(mods, Double.POSITIVE_INFINITY)

            ModUtils.applyModsToBeatmapDifficulty(difficulty, gameMode, mods, true, this@scope)

            ensureActive()

            updateThread {
                arBadge.updateStatisticBadge(selectedBeatmap.approachRate, difficulty.ar)
                odBadge.updateStatisticBadge(selectedBeatmap.overallDifficulty, difficulty.od)
                csBadge.updateStatisticBadge(selectedBeatmap.circleSize, difficulty.difficultyCS)
                hpBadge.updateStatisticBadge(selectedBeatmap.hpDrainRate, difficulty.hp)

                bpmBadge.updateStatisticBadge(
                    selectedBeatmap.mostCommonBPM.roundToInt(),
                    (selectedBeatmap.mostCommonBPM * rate).roundToInt()
                )

                scoreMultiplierBadge.updateStatisticBadge(
                    "1.00x",
                    "%.2fx".format(ModUtils.calculateScoreMultiplier(enabledMods))
                )
            }

            ensureActive()

            val attributes = when (difficultyAlgorithm) {
                droid -> calculateDroidDifficulty(beatmap, mods, this@scope)
                standard -> calculateStandardDifficulty(beatmap, mods, this@scope)
            }

            ensureActive()

            starRatingBadge.rating = attributes.starRating
            songMenu.changeDimensionInfo(selectedBeatmap)
            songMenu.setStarsDisplay(attributes.starRating.toFloat())
        }
    }

    //endregion

    //region Visibility

    override fun show() {
        GlobalManager.getInstance().engine.scene.setChildScene(
            this,
            false,
            true,
            true
        )

        // Do not show mod presets in multiplayer.
        modPresetsSection.isVisible = !Multiplayer.isMultiplayer

        if (modPresetsSection.isVisible) {
            modPresetsSection.onSearchTermUpdate(searchInput.value)
        }

        // Ensure mods and customizations that can be enabled by the user are displayed and enabled.
        updateModButtonVisibility()
        updateCustomizationMenuEnabledStates()

        // Only parsing to update mod's specific settings defaults, specially those which rely on the original beatmap data.
        parseBeatmap()
    }

    override fun back() {
        back(true)
    }

    fun back(updatePlayerMods: Boolean) {

        if (Multiplayer.isConnected) {
            Multiplayer.roomScene?.chat?.show()
            Multiplayer.roomScene?.isWaitingForModsChange = true

            // The room mods are the same as the host mods
            if (Multiplayer.isRoomHost) {
                setRoomMods(enabledMods.serializeMods())
            } else if (updatePlayerMods) {
                setPlayerMods(enabledMods.serializeMods())
            } else {
                Multiplayer.roomScene?.isWaitingForModsChange = false
            }
        }

        super.back()
    }

    //endregion

    //region Mods

    fun setMods(mods: RoomMods, isFreeMod: Boolean) {
        if (isFreeMod) {
            for ((_, mod) in enabledMods.toList()) {
                if (mod.isValidForMultiplayerAsFreeMod) {
                    continue
                }

                if (mod !in mods) {
                    removeMod(mod)
                }
            }
        } else {
            clear()
        }

        for (mod in mods.values) {
            if (!mod.isValidForMultiplayer) {
                continue
            }

            if (!isFreeMod || !mod.isValidForMultiplayerAsFreeMod) {
                addMod(mod)
            }
        }

        if (!Multiplayer.isRoomHost) {
            val doubleTime = enabledMods.ofType<ModDoubleTime>()
            val nightCore = enabledMods.ofType<ModNightCore>()

            if (Config.isUseNightcoreOnMultiplayer() && doubleTime != null) {
                removeMod(doubleTime)
                addMod(ModNightCore())
            } else if (!Config.isUseNightcoreOnMultiplayer() && nightCore != null) {
                removeMod(nightCore)
                addMod(ModDoubleTime())
            }
        }

        if (ModScoreV2::class in mods) {
            addMod(ModScoreV2())
        } else {
            removeMod(ModScoreV2())
        }

        updateModButtonVisibility()
    }

    fun updateModButtonVisibility() {
        modToggles.fastForEach {
            it.updateVisibility(searchInput.value)
            it.applyCompatibilityState()
        }

        // Update section visibility after toggles have been updated since it may not need to be visible anymore.
        modSections.fastForEach { it.updateVisibility() }
    }

    fun updateCustomizationMenuEnabledStates() {
        customizationMenu.updateComponentEnabledStates()
    }

    fun clear() {
        cancelCalculationJob()

        val room = Multiplayer.room
        val isHost = Multiplayer.isRoomHost

        enabledMods.toList().fastForEach {
            val mod = it.second

            if (room != null) {
                // For non-host in multiplayer, we want to keep mods that are not allowed to be selected by the player
                // since only the host can change those mods.
                if (!isHost && room.gameplaySettings.isFreeMod && !mod.isValidForMultiplayerAsFreeMod) {
                    return@fastForEach
                }

                // Special handling for the ScoreV2 mod, which is used by the ScoreV2 win condition.
                if (mod is ModScoreV2) {
                    return@fastForEach
                }
            }

            removeMod(mod::class)
        }
    }

    fun queueModChange(mod: Mod) = synchronized(modChangeQueue) {
        // Adding to first place in case it is already queued.
        if (modChangeQueue.isEmpty() || modChangeQueue.first() != mod) {
            if (modChangeQueue.isNotEmpty()) {
                modChangeQueue.remove(mod)
            }
            modChangeQueue.addFirst(mod)
        }
    }

    override fun onManagedUpdate(deltaTimeSec: Float) {

        if (modChangeQueue.isNotEmpty()) {
            onModsChanged()
        }

        super.onManagedUpdate(deltaTimeSec)
    }

    private fun onModsChanged() = synchronized(modChangeQueue) {
        val enabledMods = enabledMods.values.toList()
        val isRanked = enabledMods.isEmpty() || enabledMods.none { !it.isRanked }

        rankedBadge.apply {
            text = if (isRanked) "Ranked" else "Unranked"

            clearEntityModifiers()
            colorTo(if (isRanked) Color4(0xFF161622) else Theme.current.accentColor, 0.1f)

            backgroundColor = if (isRanked) Color4(0xFF83DF6B) else Theme.current.accentColor * 0.15f
        }

        modToggles.fastForEach {
            it.hasIncompatibility =
                if (!it.isSelected) enabledMods.any { m -> !it.mod.isCompatibleWith(m) } else false
        }

        scoreMultiplierBadge.updateStatisticBadge(
            "1.00x",
            "%.2fx".format(ModUtils.calculateScoreMultiplier(enabledMods))
        )

        customizeButton.isEnabled = !customizationMenu.isEmpty()

        if (modChangeQueue.any { it is IModApplicableToTrackRate }) {
            GlobalManager.getInstance().songMenu.updateMusicEffects()
        }
        modChangeQueue.clear()

        parseBeatmap()

        selectedModsIndicator.mods = enabledMods
        modPresetsSection.onModsChanged()
    }

    fun addMod(mod: Mod) {

        if (mod in enabledMods) {
            return
        }
        enabledMods.put(mod)

        modToggles.fastForEach { button ->

            val wasSelected = button.isSelected
            button.isSelected = button.mod::class in enabledMods

            if (button.mod::class == mod::class) {
                button.mod = mod
            }

            // Handle incompatible mods with the selected mod.
            if (wasSelected && !button.isSelected) {
                customizationMenu.onModRemoved(button.mod)
                button.mod = button.mod::class.createInstance()
            }
        }

        customizationMenu.onModAdded(mod)
        queueModChange(mod)
    }

    fun removeMod(mod: Mod) = removeMod(mod::class)

    fun removeMod(modClass: KClass<out Mod>) {

        if (modClass !in enabledMods) {
            return
        }

        val toggle = modToggles.find { it.mod::class == modClass } ?: return

        enabledMods.remove(modClass)

        toggle.isSelected = false

        customizationMenu.onModRemoved(toggle.mod)
        queueModChange(toggle.mod)

        toggle.mod = modClass.createInstance()
    }

    //endregion

    //region Components

    private fun <T : Comparable<T>> UILabeledBadge.updateStatisticBadge(initialValue: T, finalValue: T) {

        val newText = if (finalValue is Float || finalValue is Double) "%.2f".format(finalValue) else finalValue.toString()

        if (valueComponent.text == newText) {
            return
        }
        valueComponent.text = newText

        valueComponent.clearEntityModifiers()
        valueComponent.colorTo(Color4(when {
            initialValue < finalValue -> 0xFFF78383
            initialValue > finalValue -> 0xFF40CF5D
            else -> 0xFFFFFFFF
        }), 0.1f)
    }

    //endregion

}


